src/transform-sdk/cpp/src/transform_sdk.cc

```diff
@@ -464,8 +643,176 @@ on_record_written(const on_record_written_callback& callback) {
+        int32_t length = 0;
+        auto errc = abi::sr::get_subject_schema_len(
+          bytes_view{subject}.begin(),
+          static_cast<int32_t>(subject.size()),
+          version(),
+          &length);
+        if (errc != 0) {
+            return std::unexpected{
+              sr::make_error_code(static_cast<sr::errc>(errc))};
+        }
+        redpanda::bytes buf;
+        buf.resize(length, 0);
+        auto ec_or_n = abi::sr::get_subject_schema(
+          bytes_view{subject}.begin(),
+          static_cast<int32_t>(subject.size()),
+          version(),
+          buf.data(),
+          static_cast<int32_t>(buf.size()));
+        if (ec_or_n < 0) {
+            return std::unexpected{
+              sr::make_error_code(static_cast<sr::errc>(ec_or_n))};
+        }
+        return decode::read_schema(
+          subject, bytes_view{buf.data(), static_cast<size_t>(ec_or_n)});
+    }
+
+    [[nodiscard]] std::expected<subject_schema, std::error_code>
+    lookup_latest_schema(const std::string& subject) const final {
+        return lookup_schema_by_version(subject, schema_version{-1});
+    }
+
+    [[nodiscard]] std::expected<subject_schema, std::error_code>
+    create_schema(const std::string& subject, sr::schema schema) final {
+        bytes buf;
+        buf.reserve(schema.raw_schema().size() + varint::MAX_LENGTH);
+        decode::write_schema_def(&buf, schema);
+        sr::schema_id sid{0};
+        // NOTE: A planned ABI update will add a `version` output parameter to
+        // create_subject_schema
+        const sr::schema_version version{0};
+        auto errc = abi::sr::create_subject_schema(
+          bytes_view{subject}.begin(),
+          static_cast<int32_t>(subject.size()),
+          buf.data(),
+          static_cast<int32_t>(buf.size()),
+          sid.data());
+        if (errc != 0) {
+            return std::unexpected{
+              sr::make_error_code(static_cast<sr::errc>(errc))};
+        }
+        return sr::subject_schema{std::move(schema), subject, version, sid};
+    };
+};
+
+std::unique_ptr<schema_registry_client> new_client() {
+    return std::make_unique<abi_schema_registry_client>();
+}
+
+std::expected<std::pair<schema_id, bytes_view>, std::error_code>
+decode_schema_id(bytes_view buf) {
+    constexpr size_t HEADER_LEN = 5U;
+    static const bytes MAGIC_BYTES{0x00};
+    if (
+      buf.subview(0, MAGIC_BYTES.size()) != MAGIC_BYTES
+      || buf.size() < HEADER_LEN) {
+        return std::unexpected{
+          std::make_error_code(std::errc::illegal_byte_sequence)};
+    }
+    // NOLINTBEGIN(*-magic-numbers)
+    auto id_bytes = buf.subview(1, HEADER_LEN - 1);
+    const uint32_t raw_id = (static_cast<uint32_t>(id_bytes[3]) << 0U)
+                            | (static_cast<uint32_t>(id_bytes[2]) << 8U)
+                            | (static_cast<uint32_t>(id_bytes[1]) << 16U)
+                            | (static_cast<uint32_t>(id_bytes[0]) << 24U);
```

Comment:
Torn on casting to uint32 and using byteswap... this is probably fine :)

I always get tripped on removing UB in these things: https://justine.lol/endian.html

Reply:
ya, I think as long as there's no intermediate promotion to signed anything, we're all good. if i can avoid typing `memcpy` in these situations I'm basically happy.

Reply:
Wasm is little endian.

See list here: https://webassembly.org/docs/portability/